File: android_mobile/src/main/java/com/alexstyl/specialdates/theming/Themer.java
package com.alexstyl.specialdates.theming;

import android.content.Context;
import android.view.Window;
import android.view.WindowManager;

import com.alexstyl.specialdates.ui.base.MementoActivity;
import com.alexstyl.android.Version;

public final class Themer {

    private final ThemingPreferences preferences;
    private final AttributeExtractor attributeExtractor;

    private static Themer instance;

    public static Themer get(Context context) {
        if (instance == null) {
            ThemingPreferences preferences = ThemingPreferences.newInstance(context);
            instance = new Themer(preferences, new AttributeExtractor());
        }
        return instance;
    }

    private Themer(ThemingPreferences preferences, AttributeExtractor attributeExtractor) {
        this.preferences = preferences;
        this.attributeExtractor = attributeExtractor;
    }

    public void initialiseActivity(MementoActivity activity) {
        MementoTheme theme = preferences.getSelectedTheme();
        activity.setTheme(theme.androidTheme());

        if (Version.hasLollipop()) {
            Window window = activity.getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);

            int primaryDark = attributeExtractor.extractPrimaryColorFrom(activity);
            window.setStatusBarColor(primaryDark);
        }

    }

    public MementoTheme getCurrentTheme() {
        return preferences.getSelectedTheme();
    }

}

File: android_mobile/src/main/java/com/alexstyl/specialdates/theming/Themer.kt
package com.alexstyl.specialdates.theming

import android.app.Activity
import android.content.Context
import android.content.res.ColorStateList
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.support.annotation.DrawableRes
import android.support.v4.content.res.ResourcesCompat
import android.support.v4.graphics.drawable.DrawableCompat
import android.view.WindowManager
import com.alexstyl.android.Version

class Themer(private val preferences: ThemingPreferences,
             private val attributeExtractor: AttributeExtractor,
             private val resources: Resources) {

    fun applyThemeTo(activity: Activity) {
        val theme = preferences.selectedTheme
        activity.setTheme(theme.androidTheme())

        if (Version.hasLollipop()) {
            val window = activity.window
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)

            val primaryDark = attributeExtractor.extractPrimaryColorFrom(activity)
            window.statusBarColor = primaryDark
        }
    }

    fun getTintedDrawable(@DrawableRes drawableResId: Int, themeContext: Context): Drawable {
        val wrappedDrawable = DrawableCompat.wrap(ResourcesCompat.getDrawable(resources, drawableResId, null)!!)
        val color = attributeExtractor.extractToolbarIconColors(themeContext)
        DrawableCompat.setTintList(wrappedDrawable, ColorStateList.valueOf(color))
        return wrappedDrawable
    }
}
